drivers/mbox/mbox_nrf_bellboard_local.c

```diff
@@ -0,0 +1,169 @@
+			nrf_bellboard_event_clear(bellboard, event);
+		}
+
+		if ((int_pend & BIT(i)) != 0U) {
+			if (cbs[i] != NULL) {
+				cbs[i](DEVICE_DT_INST_GET(0), i, cbs_ctx[i], NULL);
+			}
+		}
+	}
+}
+
+static uint32_t bellboard_local_max_channels_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT;
+}
+
+static int bellboard_local_register_callback(const struct device *dev, uint32_t id,
+					     mbox_callback_t cb, void *user_data)
+{
+	ARG_UNUSED(dev);
+
+	if (id >= NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT) {
+		return -EINVAL;
+	}
+
+	cbs[id] = cb;
+	cbs_ctx[id] = user_data;
+
+	return 0;
+}
+
+static int bellboard_local_set_enabled(const struct device *dev, uint32_t id, bool enable)
+{
+	bool valid_found = false;
+
+	ARG_UNUSED(dev);
+
+	if (id >= NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT) {
+		return -EINVAL;
+	}
+
+	for (uint8_t i = 0U; i < BELLBOARD_NUM_IRQS; i++) {
+		uint32_t *evt_enabled_mask;
+
+		if ((evt_mappings[i] == 0U) || ((evt_mappings[i] & BIT(id)) == 0U)) {
+			continue;
+		}
+
+		valid_found = true;
+		evt_enabled_mask = &evt_enabled_masks[i];
+
+		if (enable) {
+			if ((*evt_enabled_mask & BIT(id)) != 0U) {
+				return -EALREADY;
+			}
+
+			*evt_enabled_mask |= BIT(id);
+			nrf_bellboard_int_enable(bellboard, i, BIT(id));
+		} else {
+			if ((*evt_enabled_mask & BIT(id)) == 0U) {
+				return -EALREADY;
+			}
+
+			*evt_enabled_mask &= ~BIT(id);
+			nrf_bellboard_int_disable(bellboard, i, BIT(id));
+		}
+	}
+
+	if (!valid_found) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct mbox_driver_api bellboard_local_driver_api = {
+	.max_channels_get = bellboard_local_max_channels_get,
+	.register_callback = bellboard_local_register_callback,
+	.set_enabled = bellboard_local_set_enabled,
+};
+
+#define BELLBOARD_IRQ_CONFIGURE(name, idx)                                                         \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, name),                                                 \
+		    (IRQ_CONNECT(DT_INST_IRQ_BY_NAME(0, name, irq),                                \
+				 DT_INST_IRQ_BY_NAME(0, name, priority), bellboard_local_isr,      \
+				 (const void *)idx, 0);                                            \
+		     irq_enable(DT_INST_IRQ_BY_NAME(0, name, irq));),                              \
+		    ())
+
+static int bellboard_local_init(const struct device *dev)
+{
+	uint32_t evt_all_mappings =
+		evt_mappings[0] | evt_mappings[1] | evt_mappings[2] | evt_mappings[3];
+
+	ARG_UNUSED(dev);
+
+	nrf_bellboard_int_disable(bellboard, 0, evt_mappings[0]);
+	nrf_bellboard_int_disable(bellboard, 1, evt_mappings[1]);
+	nrf_bellboard_int_disable(bellboard, 2, evt_mappings[2]);
+	nrf_bellboard_int_disable(bellboard, 3, evt_mappings[3]);
+
+	for (uint8_t i = 0U; i < NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT; i++) {
+		if ((evt_all_mappings & BIT(i)) != 0U) {
+			nrf_bellboard_event_clear(bellboard, nrf_bellboard_triggered_event_get(i));
```

Comment:
I don't think the driver should clear all the events on initialization.
We can imagine that a CPU X is in a SoftOff state and it is woken up by a BELLBOARD event by another CPU Y. The event is pending. Then the CPU X runs a complete cold-boot procedure, including the bellboard driver initialization. After the initialization is complete and the application code enables bellboard events and IRQs, the pending event must be handled. If the event gets cleared during the driver initialization, the event will be lost and CPU X won't have any idea why it was woken up.

Reply:
Will drop this. Note that I inherited this code from nrfy, so it has always been done like that. I guess it wasn’t the right thing to do.

Reply:
Note: I'll open a follow-up PR resolving any remaining issues after we have a call to discuss more details, since the logic in this PR has been largely based on what we already had before. Also, it's already part of sdk-nrf in its current form (and it works "well enough" to unblock certain work), so we do not have the same pressure as before to discuss technical aspects as before :-)

Reply:
What I am afraid of is that scheme of clearing events in nrfx drivers and nrfy layer during initialization is established for quite some time. Rationale is that during initialization driver applies clear & known hardware configuration, to avoid situation in which leftover peripheral register contents alter the driver behavior.

If we decide to change the nrfx behavior in this manner (that is to handle events that might be triggered before driver initialization and shall be considered as valid) IMO such change should be done systematically across all drivers, instead of doing it ad-hoc here

Reply:
IPC signaling peripherals might be a special case requiring a different approach. But let's discuss it in the "follow up PR" scope mentioned by @gmarull